azure/src/main/scala/quasar/blobstore/azure/handlers.scala

```diff
@@ -19,50 +19,77 @@ package quasar.blobstore.azure
 import quasar.blobstore.BlobstoreStatus
 
 import java.lang.Throwable
-
-import scala.{Byte, Option}
+import java.nio.ByteBuffer
+import scala.{Byte, Int, None, Option, Some, StringContext, Unit}
 import scala.util.control.NonFatal
 
 import cats.ApplicativeError
 import cats.data.Kleisli
-import cats.effect.ConcurrentEffect
+import cats.effect.Sync
 import cats.instances.int._
+import cats.syntax.applicative._
 import cats.syntax.applicativeError._
+import cats.syntax.apply._
 import cats.syntax.eq._
 import cats.syntax.functor._
 import cats.syntax.option._
-import com.microsoft.azure.storage.blob.{DownloadResponse, ReliableDownloadOptions, StorageException}
+import com.azure.storage.blob.models.BlobStorageException
 import fs2.{Chunk, Stream}
 
 object handlers {
 
-  def recoverToBlobstoreStatus[F[_]: ApplicativeError[?[_], Throwable], A](fa: F[A]): F[BlobstoreStatus] =
-    fa.map(_ => BlobstoreStatus.ok()).recover {
-      case ex: StorageException if ex.statusCode === 403 => BlobstoreStatus.noAccess()
-      case ex: StorageException if ex.statusCode === 404 => BlobstoreStatus.notFound()
-      case ex: StorageException => BlobstoreStatus.notOk(ex.message())
+  def recoverToBlobstoreStatus[F[_]: ApplicativeError[?[_], Throwable]](fa: F[BlobstoreStatus]): F[BlobstoreStatus] =
+    fa.recover {
+      case ex: BlobStorageException if ex.getStatusCode === 403 =>
+        BlobstoreStatus.noAccess()
+      case ex: BlobStorageException if ex.getStatusCode === 404 =>
+        BlobstoreStatus.notFound()
+      case ex: BlobStorageException =>
+        BlobstoreStatus.notOk(s"Status: ${ex.getStatusCode} Message:${ex.getMessage}")
       case NonFatal(t) => BlobstoreStatus.notOk(t.getMessage)
     }
 
-  def recoverNotFound[F[_], A](fa: F[A])(implicit F: ApplicativeError[F, Throwable]): F[Option[A]] =
+  def recoverToStatusCode[F[_]: ApplicativeError[?[_], Throwable]](fa: F[Int]): F[Int] =
+    fa.recover {
+      case ex: BlobStorageException => ex.getStatusCode()
+      case NonFatal(_) => 500
+    }
+
+  def recoverToNone[F[_], A](fa: F[A])(implicit F: ApplicativeError[F, Throwable]): F[Option[A]] =
     F.recover(fa.map(_.some)) {
-      case ex: StorageException if ex.statusCode() === 404 => none
+      case NonFatal(_) => None
     }
 
+  def raiseInnerStreamError[F[_]: Sync, A](
+      s: Stream[F,A])
+      : F[Stream[F, A]] =
+    // we are only interested in errors at the beginning of the stream
+    // s.take(1) is sufficient rather than the whole stream s
+    s.take(1).compile.drain.handleErrorWith(_.raiseError[F, Unit]) *> s.pure[F]
```

Comment:
This re-runs the stream (with all the effects and resource acquisition that entails) when the user ultimately reads it. We *definitely* don't want to do this. Why exactly are you trying to re-sequence the error here? And can it be re-sequenced at the point at which it happens, rather than trying to recover it from the stream after the fact?

Reply:
@djspiewak Hmm, I know there was negative performance impact with this construction but
I thought with the `take(1)` that was largely addressed. See my comments here https://github.com/precog/async-blobstore/pull/96#discussion_r426296844 .

I still like to understand this better, but I can implement the isExisting check that I mentioned in that comment, then this function should not be necessary at all. 
Or if you have a better idea to lift the inner stream errors up, that would be even better because I have been struggling with this for a while.